drv/lpc55-swd/src/main.rs

```diff
@@ -0,0 +1,656 @@
+    }
+}
+
+// The parity is only over 4 of the bits
+fn calc_parity(val: u8) -> u8 {
+    let b = val >> 3 & 0xf;
+
+    if b.count_ones() % 2 == 0 {
+        0 << PARITY_BIT
+    } else {
+        1 << PARITY_BIT
+    }
+}
+
+struct ServerImpl {
+    spi: spi_core::Spi,
+    init: bool,
+}
+
+impl idl::InOrderSpCtrlImpl for ServerImpl {
+    fn read(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        dest: LenLimit<Leased<W, [u8]>, 4096>,
+    ) -> Result<(), RequestError<SpCtrlError>> {
+        ringbuf_entry!(Trace::ReadCmd);
+        if !self.init {
+            return Err(SpCtrlError::NeedInit.into());
+        }
+        let cnt = dest.len();
+        if cnt % 4 != 0 {
+            return Err(SpCtrlError::BadLen.into());
+        }
+        let mut buf = LeaseBufWriter::<_, 32>::from(dest.into_inner());
+
+        for i in 0..cnt / 4 {
+            match self.read_target_addr(addr + ((i * 4) as u32)) {
+                Ok(r) => {
+                    ringbuf_entry!(Trace::MemVal(r));
+                    for b in r.to_le_bytes() {
+                        if buf.write(b).is_err() {
+                            return Ok(());
+                        }
+                    }
+                }
+                Err(_) => return Err(SpCtrlError::Fault.into()),
+            }
+        }
+
+        Ok(())
+    }
+
+    fn write(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        dest: LenLimit<Leased<R, [u8]>, 4096>,
+    ) -> Result<(), RequestError<SpCtrlError>> {
+        ringbuf_entry!(Trace::WriteCmd);
+        if !self.init {
+            return Err(SpCtrlError::NeedInit.into());
+        }
+        let cnt = dest.len();
+        if cnt % 4 != 0 {
+            return Err(SpCtrlError::BadLen.into());
+        }
+        let mut buf = LeaseBufReader::<_, 32>::from(dest.into_inner());
+
+        for i in 0..cnt / 4 {
+            let mut word: [u8; 4] = [0; 4];
+            for j in 0..4 {
+                match buf.read() {
+                    Some(b) => word[j] = b,
+                    None => return Ok(()),
+                };
+            }
+            match self.write_target_addr(
+                addr + ((i * 4) as u32),
+                u32::from_le_bytes(word),
+            ) {
+                Err(_) => return Err(SpCtrlError::Fault.into()),
+                _ => (),
+            }
+        }
+
+        Ok(())
+    }
+
+    fn setup(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError<SpCtrlError>> {
+        match self.swd_setup() {
+            Ok(_) => {
+                self.init = true;
+                Ok(())
+            }
+            Err(_) => Err(SpCtrlError::Fault.into()),
+        }
+    }
+}
+
+impl ServerImpl {
+    fn io_out(&mut self) {
+        while !self.spi.mstidle() {
+            cortex_m::asm::nop();
+        }
+
+        let gpio_driver = GPIO.get_task_id();
```

Comment:
Rather than making a syscall every time to look up the current generation, I recommend creating the `Pins` driver facade _once_ and storing it as a field in your server. It will update its own stored notion of the generation on any failure. This should save text and time.